### ISSUE TITLE
Out-of-bounds read in CDFContextLogOps::push

### ISSUE THREAD

**Issue by CodesInChaos** [xiph/rav1e#3166]
`CDFContextLogOps::push` looks like this:

```rust
  fn push(log: &mut CDFContextLogBase, cdf: &[u16]) {
    debug_assert!(cdf.len() <= Self::CDF_LEN_MAX);
    let offset = cdf.as_ptr() as usize - log.base as usize;
    debug_assert!(offset <= u16::MAX.into());
    // SAFETY: Maintain an invariant of non-zero spare capacity, so that
    // branching may be deferred until writes are issued. Benchmarks indicate
    // this is faster than first testing capacity and possibly reallocating.
    unsafe {
      let len = log.data.len();
      let new_len = len + Self::CDF_LEN_MAX + 1;
      let capacity = log.data.capacity();
      debug_assert!(new_len <= capacity);
      let dst = log.data.as_mut_ptr().add(len);
      dst.copy_from_nonoverlapping(cdf.as_ptr(), Self::CDF_LEN_MAX);
      *dst.add(Self::CDF_LEN_MAX) = offset as u16;
      log.data.set_len(new_len);
      if Self::CDF_LEN_MAX + 1 > capacity.wrapping_sub(new_len) {
        log.data.reserve(Self::CDF_LEN_MAX + 1);
      }
    }
  }
```

The problematic line is

```rust
dst.copy_from_nonoverlapping(cdf.as_ptr(), Self::CDF_LEN_MAX);
```

`copy_from_nonoverlapping` requires that `cdf.len() >= Self::CDF_LEN_MAX`. If I add a debug assertion verifying this precondition, it fails:

```rust
debug_assert!(cdf.len() >= Self::CDF_LEN_MAX);
dst.copy_from_nonoverlapping(cdf.as_ptr(), Self::CDF_LEN_MAX);
```

This means that the read is out-of-bounds, and thus undefined behaviour.

**Comment by barrbrain** [comment]
This is not an over-read bug. The contract is that CDFs belong to a larger data structure that guarantees reading past the end of  a CDF is well-defined. However, documenting that with the right assertions is most welcome.

**Comment by CodesInChaos** [comment]
I believe still is still illegal in the stacked-borrows memory model, since the pointer is derived from the shorter slice and thus does not have the right to access memory outside the slice. Though it's not entirely clear if stacked-borrows will become rust's memory model once it's formalized.

Consider this piece of code:

```rust
let src_buf = [0u8; 32];
let src_slice = &src_buf[0..16];
let src_ptr = src_slice.as_ptr();

let mut dst_buf = [0u8; 32];
let dst_ptr = dst_buf.as_mut_ptr();
unsafe { dst_ptr.copy_from_nonoverlapping(src_ptr, 32) };
```

Here `src_ptr` is derived from a slice that's too short (len=16) while it's part of a buffer `src_buf` which is long enough (len=32). However miri does claim undefined behaviour in this code [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0c5a229ed24199cf9055bee3d452da07):

```
error: Undefined Behavior: attempting a read access using <2937> at alloc1535[0x10], but that tag does not exist in the borrow stack for this location
 --> src/main.rs:8:14
  |
8 |     unsafe { dst_ptr.copy_from_nonoverlapping(src_ptr, 32) };
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |              |
  |              attempting a read access using <2937> at alloc1535[0x10], but that tag does not exist in the borrow stack for this location
  |              this error occurs as part of an access at alloc1535[0x0..0x20]
  |
  = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
  = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <2937> was created by a SharedReadOnly retag at offsets [0x0..0x10]
 --> src/main.rs:4:19
  |
4 |     let src_ptr = src_slice.as_ptr();
  |                   ^^^^^^^^^^^^^^^^^^
```

**Comment by barrbrain** [comment]
Thank you for the wider context of Rust memory models. It would be great if we could express this feature in way that aligns with an existing model. My intuition is that we would have to remove the use of slices as a convenient way to anonymize CDFs. I think we may need to introduce an operator on the larger data structure that achieves the same result.

**Comment by barrbrain** [comment]
I have verified with miri that the access pattern is compliant with the Stacked Borrows model after #3172 was merged.
